js/editor-controller.js

```diff
@@ -0,0 +1,235 @@
+                        }
+                        if (event.data.projects.length < 1) {
+                            throw new Error("'projects' array should contain at least one item.");
+                        }
+                        _editorActions.setCode(event.data.projects[0]);
+                        // Notify parent about editor successfully configured
+                        window.parent.postMessage({
+                            type: CONTROLLER_MESSAGING.type,
+                            action: CONTROLLER_MESSAGING.actions.workspaceloaded
+                        }, '*');
+                        break;
+
+                    // Parent is sending HEX file
+                    case CONTROLLER_MESSAGING.actions.loadhex:
+                        _editorActions.loadHex(event.data.filename, event.data.hexstring);
+                        break;
+
+                    // Parent is sending file for filesystem
+                    case CONTROLLER_MESSAGING.actions.loadfile:
+                        _editorActions.loadFileToFs(event.data.filename, event.data.filestring);
+                        break;
+
+                    // Parent is requesting postMessage downloads
+                    case CONTROLLER_MESSAGING.actions.downloadmode:
+                        usePostMessageFlashing = event.data.enabled;
+                        break;
+
+                    default:
+                        throw new Error('Unsupported action.')
+                }
+            }
+        }, false);
+    };
+
+    /**
+     * Initialises the controller based on the URL query string parameters and
+     * the `window` environment.
+     * The Classroom controller will be an iframe parent, and the mobile apps
+     * will have postMessage handlers on different namespaces within `window`.
+     *
+     * @param {object} qs - The editor URL query string parameters.
+     */
+    _returnObj.initialise = function(qs) {
+        var inIframe = window !== window.parent;
+        var iframeControllerMode = inIframe && qs.controller == '1';
+        var appControllerMode = qs.mobileApp == '1';
+
+        if (iframeControllerMode || appControllerMode) {
+            // Detect the host controller to send our messages
+            if (iframeControllerMode && window.parent && window.parent.postMessage) {
+                // Classroom wraps the editor in an iframe
+                _controllerHost = window.parent;
+            } else if (appControllerMode &&
+                    window.webkit &&
+                    window.webkit.messageHandlers &&
+                    window.webkit.messageHandlers.host &&
+                    window.webkit.messageHandlers.host.postMessage) {
+                // iOS app, 'host' used to match the MakeCode messageHandler name
+                _controllerHost = window.webkit.messageHandlers.host;
+            } else if (appControllerMode && window.host && window.host.postMessage) {
```

Comment:
I believe we (the apps) have full control of defining these, correct?

My original suggestion of adding the `downloadmode` was because I thought maybe these namespaces were android/ios/browser/webview dependant and we could get false positives.

Reply:
The name "host" can be anything. The iOS app can register a message handler with the chosen name, which is I guess when window.webkit.messageHandlers.whatever becomes defined for javascript. I'm not aware of the `downloadmode` suggestion.